[WebXR] Allow audio and media sessions to continue uninterrupted when starting immersive WebXR sessions

achan00 opened the pull request:
#### 76e52d168312e91a83bd2f9db3535f6921b3cd87
<pre>
[WebXR] Allow audio and media sessions to continue uninterrupted when starting immersive WebXR sessions
<a href="https://bugs.webkit.org/show_bug.cgi?id=269332">https://bugs.webkit.org/show_bug.cgi?id=269332</a>
<a href="https://rdar.apple.com/102391109">rdar://102391109</a>

Reviewed by Eric Carlson.

The original Safari window gets backgrounded when we start an immersive WebXR
session. We need to allow audio/media to continue playback when we get a
backgrounding notification for this scenario, similar to how we handle this
for picture-in-picture.

* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::shouldOverrideBackgroundPlaybackRestriction const):
Return true if the page has an active WebXR immersive session.
* Source/WebCore/Modules/webxr/WebXRSystem.cpp:
(WebCore::WebXRSystem::hasActiveImmersiveSession const):
* Source/WebCore/Modules/webxr/WebXRSystem.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::shouldOverrideBackgroundPlaybackRestriction const):
Return true if the page has an active WebXR immersive session.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::hasActiveImmersiveSession const):
Walk the frame tree and return true if any frame&apos;s window has an active
WebXR immersive session.
* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateDrawingAreaLayerTreeFreezeState):
Similar to picture-in-picture, skip freezing the layer tree state if
the window gets backgrounded due to an active immersive session.

Canonical link: <a href="https://commits.webkit.org/274742@main">https://commits.webkit.org/274742@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/624e28fce057a2fde114920acae1c2f9b33eec9a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39594 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18573 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41950 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42129 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35494 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41900 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21463 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15902 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33059 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40168 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15657 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34267 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13580 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13559 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35220 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43406 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35947 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35551 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39343 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14406 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11864 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37610 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16012 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8951 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16060 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15669 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -8570,6 +8570,14 @@ bool HTMLMediaElement::shouldOverrideBackgroundPlaybackRestriction(PlatformMedia
         }
 #endif
+#if PLATFORM(VISION) && ENABLE(WEBXR)
+        if (Page* page = document().page()) {
```
```suggestion
        if (RefPtr page = document().page()) {
```

pxlcoder commented on Source/WebCore/Modules/webaudio/AudioContext.cpp:
```diff
@@ -492,7 +496,19 @@ void AudioContext::isPlayingAudioDidChange()
+
+#if PLATFORM(VISION) && ENABLE(WEBXR)
+    RefPtr document = this->document();
+    if (document) {
```
```suggestion
    if (RefPtr document = this->document()) {
```

pxlcoder commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4670,4 +4675,24 @@ void Page::setAccessibilityRootObject(AccessibilityRootAtspi* rootObject)
+#if PLATFORM(IOS_FAMILY) && ENABLE(WEBXR)
+bool Page::hasActiveImmersiveSession() const
+{
+    for (auto* frame = &mainFrame(); frame; frame = frame->tree().traverseNext()) {
```
```suggestion
    for (RefPtr frame = &mainFrame(); frame; frame = frame->tree().traverseNext()) {
```

pxlcoder replied:
Please use `RefPtr` below where possible too.

achan00 committed:
[WebXR] Allow audio and media sessions to continue uninterrupted when starting immersive WebXR sessions
https://bugs.webkit.org/show_bug.cgi?id=269332
rdar://102391109

Reviewed by Eric Carlson.

The original Safari window gets backgrounded when we start an immersive WebXR
session. We need to allow audio/media to continue playback when we get a
backgrounding notification for this scenario, similar to how we handle this
for picture-in-picture.

* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::shouldOverrideBackgroundPlaybackRestriction const):
Return true if the page has an active WebXR immersive session.
* Source/WebCore/Modules/webxr/WebXRSystem.cpp:
(WebCore::WebXRSystem::hasActiveImmersiveSession const):
* Source/WebCore/Modules/webxr/WebXRSystem.h:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::shouldOverrideBackgroundPlaybackRestriction const):
Return true if the page has an active WebXR immersive session.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::hasActiveImmersiveSession const):
Walk the frame tree and return true if any frame's window has an active
WebXR immersive session.
* Source/WebCore/page/Page.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateDrawingAreaLayerTreeFreezeState):
Similar to picture-in-picture, skip freezing the layer tree state if
the window gets backgrounded due to an active immersive session.

Canonical link: https://commits.webkit.org/274742@main
